### PR TITLE
Map enhancement

### PR DESCRIPTION
**Pull request by TheCloudlet**
The enhancement information as GitHub page - [rb_bench](https://github.com/ypaskell/rbtree_bench).
Correctness tested is provided.
Issue #29

### CODE REVIEW

**Comment by ypaskell** [comment]
Hi @jserv, all CodeQL defects has been resolved.

**Review by jserv — CHANGES_REQUESTED** [review]
Squash the commits and emphasize the benefits of this proposed change.

**Comment by jserv** [comment]
You shall utilize `__UNREACHABLE` when possible. See https://github.com/sysprog21/rv32emu/blob/master/src/common.h

File: src/map.h [hunk 1]
```diff
@@ -3,70 +3,60 @@
...
+ * bit)
  *
  * The red-black tree consists of a root and nodes attached to this root.
  */
 typedef struct map_node {
     void *key, *data;
+    struct map_node *left, *right_red; /* red-black tree */
+} map_node_t;
+
+/* TODO: Change to binary representation and utilize a pointer to the node
```
Comment (jserv) [inline src/map.h]: This item MUST be completed.

**Comment by jserv** [comment]
Let's synchronize with the implementation in [rbtree_bench](https://github.com/ypaskell/rbtree_bench).

**Comment by TheCloudlet** [comment]
I may need to rewrite a good commit message.
Need some time.

**Review by jserv — CHANGES_REQUESTED** [review]
Exclude the generated `map.h.gch` (pre-compiled header).

File: src/map.h [hunk 2]
```diff
@@ -3,70 +3,54 @@
  * "LICENSE" for information on usage and redistribution of this file.
  */
 
-/*
```
Comment (jserv) [inline src/map.h]: Don't remove this informative comment.

File: src/map.c [hunk 3]
```diff
@@ -3,752 +3,641 @@
...
- * @color: new color of the node
- */
-static inline void rb_set_parent_color(map_node_t *node,
-                                       map_node_t *parent,
-                                       map_color_t color)
+/* Right accessors */
+static inline map_node_t *rb_node_get_right(const map_node_t *node)
 {
-    node->parent_color = (unsigned long) parent | color;
+    return (map_node_t *) (((uintptr_t) node->right_red) & ~3);
```
Comment (jserv) [inline src/map.c:58]: Evaluate https://github.com/torvalds/linux/commit/b0687c1119b4e8c88a651b6e876b7eae28d076e3 and integrate into this pull request (as another commit).

**Comment by jserv** [comment]
> I may need to rewrite a good commit message. Need some time.

Useful prompts for ChatGPT:
* Proofread the git commit message "..."
* Write in third-person point of view
* Wrap the body of git commit messages at 72 characters

File: src/map.c [hunk 4]
```diff
@@ -3,752 +3,641 @@
...
+        map_cmp_t cmp = pathp->cmp = (rb->cmp)(node->key, pathp->node->key);
+        switch (cmp) {
+        case _CMP_LESS:
+            pathp[1].node = rb_node_get_left(pathp->node);
+            break;
+        case _CMP_GREATER:
+            pathp[1].node = rb_node_get_right(pathp->node);
+            break;
+        default:
+            /* igore duplicate key */
```
Comment (jserv) [inline src/map.c:159]: Shall we add a proper `__UNREACHABLE` here?

Comment (TheCloudlet) [reply src/map.c:159]: Yes, after consideration, we should add this. And I have added.

File: src/map.c (hunk 3 above)
Comment (TheCloudlet) [reply src/map.c:58]: I do not think it is a good idea.

In linux the usage `rb_set_black` is only used to set node from red to black.
However, the map `rb_node_set_red` we implemented can set node from any state to black.
If we change the interface, we will cause a lot of unneeded branch that hurts the performance.

File: src/map.c [hunk 5]
```diff
@@ -3,752 +3,642 @@
...
 
-    /* Properties */
-    size_t key_size, element_size, size;
+    /* properties */
+    size_t key_size, data_size;
 
-    map_iter_t it_end, it_most, it_least;
-
-    int (*comparator)(const void *, const void *);
+    map_cmp_t (*cmp)(const void *, const void *);
```
Comment (jserv) [inline src/map.c]: Rename it to `comparator` to emphasize its action.

File: src/map.c [hunk 6]
```diff
@@ -3,752 +3,642 @@
...
+    }
+    pathp->node = node;
 
-    /* Refigure out the identity */
-    node = parent->right;
-    grandparent = rb_parent(parent);
-    uncle =
-        (grandparent->left == parent) ? grandparent->right : grandparent->left;
+    /* The loop invariant maintained is that all nodes with out-of-date
+     * summaries live in path[0], path[1], ..., *pathp.
```
Comment (jserv) [inline src/map.c]: Refine the comment. There is no "summary" in current implementation.

File: src/map.c [hunk 7]
```diff
@@ -3,752 +3,642 @@
...
+     * To maintain this, summarizing the node is necessary, as pathp is
+     * decremented before the first iteration.
+     */
+    assert(!rb_node_get_left(node));
+    assert(!rb_node_get_right(node));
+    /* Unwind.
```
Comment (jserv) [inline src/map.c]: Ditto. Improve the comment.

File: src/map.c [hunk 8]
```diff
@@ -3,752 +3,632 @@
...
 #include <stdint.h>
 #include <stdlib.h>
 #include <string.h>
 
 #include "map.h"
 
-struct map_internal {
-    struct map_node *head;
+/* TODO: Avoid relying on key_size and data_size */
+struct map_head_t {
```
Comment (jserv) [inline src/map.c]: The suffix `_t` is used for typedef. Refine the naming. I think `map_head` is confusing.

Comment (ypaskell) [reply src/map.c:24]: How about `rb_entry`?

**Review by jserv — CHANGES_REQUESTED** [review]
Revise the git commit message as [How to Write a Git Commit Message](https://cbea.ms/git-commit/) suggested.
1. [Do not end the subject line with a period](https://cbea.ms/git-commit/#end)
2. [Use the body to explain what and why vs. how](https://cbea.ms/git-commit/#why-not-how)

You shall describe the experimental environment where you perform.

**Comment by jserv** [comment]
> Revise the git commit message as [How to Write a Git Commit Message](https://cbea.ms/git-commit/) suggested.

Append the folowing:
> This map implementation has undergone extensive modifications, heavily relying on the rb.h header file from jemalloc.
> The original rb.h file served as the foundation and source of inspiration for adapting and tailoring it specifically for this map implementation.
> Therefore, credit and sincere thanks are extended to jemalloc for their invaluable work.
> Reference: https://github.com/jemalloc/jemalloc/blob/dev/include/jemalloc/internal/rb.h

File: src/map.c [hunk 9]
```diff
@@ -3,752 +3,632 @@
...
 #include <stdint.h>
 #include <stdlib.h>
 #include <string.h>
 
 #include "map.h"
 
-struct map_internal {
-    struct map_node *head;
+/* TODO: Avoid relying on key_size and data_size */
+struct rb_entry_point {
```
Comment (jserv) [inline src/map.c]: Name it `map_internal`.

File: src/map.c [hunk 10]
```diff
@@ -3,752 +3,632 @@
...
-        parent = *indirect;
-        indirect = res < 0 ? &(*indirect)->left : &(*indirect)->right;
-    }
-
-    *indirect = new_node;
-    rb_set_parent(new_node, parent);
-    map_fix_colors(obj, new_node);
-    map_calibrate(obj);
-    return true;
+    map_t tree = malloc(sizeof(struct rb_entry_point));
```
Comment (jserv) [inline src/map.c]: Add `assert` to validate `tree`.

**Comment by jserv** [comment]
Thank @ypaskell for contributing!

**Pull request merged**